src/category_theory/monoidal/Bimod.lean

```diff
@@ -0,0 +1,1043 @@
+begin
+  refine coequalizer.desc (inv_hom_aux P Q L) _,
+  dunfold inv_hom_aux, dsimp,
+  refine (@cancel_epi _ _ _ _ _ (𝟙 _ ⊗ coequalizer.π _ _) (map_π_epi (tensor_left _) _ _) _ _).1 _,
+  dunfold tensor_Bimod.X, dsimp,
+  slice_lhs 1 2 { rw [id_tensor_comp_tensor_id, ←tensor_id_comp_id_tensor] },
+  slice_lhs 2 4 { rw id_tensor_π_preserves_coequalizer_inv_desc },
+  slice_lhs 1 2 { rw [←monoidal_category.tensor_id, associator_inv_naturality] },
+  slice_lhs 2 3 { rw [←comp_tensor_id, coequalizer.condition, comp_tensor_id, comp_tensor_id] },
+  slice_rhs 1 2 { rw [←monoidal_category.tensor_id, associator_naturality] },
+  slice_rhs 2 3 { rw [←id_tensor_comp,
+                      tensor_Bimod.id_tensor_π_act_left,
+                      id_tensor_comp, id_tensor_comp] },
+  slice_rhs 4 6 { rw id_tensor_π_preserves_coequalizer_inv_desc },
+  slice_rhs 3 4 { rw associator_inv_naturality },
+  coherence,
+end
+
+lemma hom_hom_inv_hom_id : hom_hom P Q L ≫ inv_hom P Q L = 𝟙 _ :=
+begin
+  dunfold hom_hom hom_hom_aux inv_hom inv_hom_aux, dsimp,
+  ext,
+  slice_lhs 1 2 { rw coequalizer.π_desc },
+  refine (@cancel_epi _ _ _ _ _ (coequalizer.π _ _ ⊗ 𝟙 _) (map_π_epi (tensor_right _) _ _) _ _).1 _,
+  slice_lhs 1 3 { rw π_tensor_id_preserves_coequalizer_inv_desc },
+  slice_lhs 3 4 { rw coequalizer.π_desc },
+  slice_lhs 2 4 { rw id_tensor_π_preserves_coequalizer_inv_desc },
+  slice_lhs 1 3 { rw iso.hom_inv_id_assoc },
+  dunfold tensor_Bimod.X,
+  slice_rhs 2 3 { rw category.comp_id },
+  refl,
+end
+
+lemma inv_hom_hom_hom_id : inv_hom P Q L ≫ hom_hom P Q L = 𝟙 _ :=
+begin
+  dunfold hom_hom hom_hom_aux inv_hom inv_hom_aux, dsimp,
+  ext,
+  slice_lhs 1 2 { rw coequalizer.π_desc },
+  refine (@cancel_epi _ _ _ _ _ (𝟙 _ ⊗ coequalizer.π _ _) (map_π_epi (tensor_left _) _ _) _ _).1 _,
+  slice_lhs 1 3 { rw id_tensor_π_preserves_coequalizer_inv_desc },
+  slice_lhs 3 4 { rw coequalizer.π_desc },
+  slice_lhs 2 4 { rw π_tensor_id_preserves_coequalizer_inv_desc },
+  slice_lhs 1 3 { rw iso.inv_hom_id_assoc },
+  dunfold tensor_Bimod.X,
+  slice_rhs 2 3 { rw category.comp_id },
+  refl,
+end
+
+end associator_Bimod
+
+noncomputable
+def associator_Bimod {W X Y Z : Mon_ C} (L : Bimod W X) (M : Bimod X Y) (N : Bimod Y Z) :
+  tensor_Bimod (tensor_Bimod L M) N ≅ tensor_Bimod L (tensor_Bimod M N) :=
+iso_of_iso
+  { hom := associator_Bimod.hom_hom L M N,
+    inv := associator_Bimod.inv_hom L M N,
+    hom_inv_id' := associator_Bimod.hom_hom_inv_hom_id L M N,
+    inv_hom_id' := associator_Bimod.inv_hom_hom_hom_id L M N }
+  (associator_Bimod.hom_left_act_hom' L M N)
+  (associator_Bimod.hom_right_act_hom' L M N)
+
+namespace left_unitor_Bimod
+variables {R S : Mon_ C} (P : Bimod R S)
+
+noncomputable
+def hom_hom : tensor_Bimod.X (regular R) P ⟶ P.X :=
+begin
+  dsimp, dunfold tensor_Bimod.X, dsimp,
+  refine coequalizer.desc P.act_left _,
+  rw [category.assoc, left_assoc],
+end
```

Comment:
By the way, one thing I deliberately tried to avoid is unconstrained `simp`s because they seem to make everything really slow.  That's why some of the proofs are longer than they could be.